src/api/ucc.h

```diff
@@ -0,0 +1,709 @@
+ * attributes. The attributes of the context handle are described by the context
+ * attributes @ref ucc_context_attrib_t
+ *
+ *  @param [in]   context          Communication context
+ *  @param [out]  context_attrib   Attributes of the communication context
+ * 
+ *  @return Error code as defined by @ref ucc_status_t
+ */
+
+ucc_status_t ucc_context_get_attribs(ucc_context_h context, 
+                                     ucc_context_attribs_t *context_atrib);
+
+
+/**
+ * *************************************************************
+ *                   Teams Section 
+ * *************************************************************
+ */
+
+enum ucc_team_params_field {
+    UCC_PARAMS_POST_ORDERING          = UCC_BIT(0),
+    UCC_PARAMS_OUTSTANDING_CALLS      = UCC_BIT(1),
+    UCC_PARAMS_EP                     = UCC_BIT(2),
+    UCC_PARAMS_EP_TYPE                = UCC_BIT(3),
+    UCC_PARAMS_SYNC_TYPE              = UCC_BIT(4),
+    UCC_PARAMS_OOB                    = UCC_BIT(5),
+    UCC_PARAMS_MEM_PARAMS             = UCC_BIT(6)
+};
+
+enum ucc_team_attribs_field {
+    UCC_ATTRIBS_POST_ORDERING          = UCC_BIT(0),
+    UCC_ATTRIBS_OUTSTANDING_CALLS      = UCC_BIT(1),
+    UCC_ATTRIBS_EP                     = UCC_BIT(2),
+    UCC_ATTRIBS_EP_TYPE                = UCC_BIT(3),
+    UCC_ATTRIBS_SYNC_TYPE              = UCC_BIT(4),
+    UCC_ATTRIBS_OOB                    = UCC_BIT(5),
+    UCC_MEM_PARAMS                     = UCC_BIT(6)
+};
+
+typedef enum  {
+    UCC_MEM_SYMMETRIC   = UCC_BIT(0),
+    UCC_MEM_PERSISTENT  = UCC_BIT(1),
+    UCC_MEM_ALIGN32     = UCC_BIT(2),
+    UCC_MEM_ALIGN64     = UCC_BIT(3),
+    UCC_MEM_ALIGN128    = UCC_BIT(4),
+} ucc_mem_constraints_t;
+
+typedef enum {
+	UCC_REMOTE_ATOMICS,
+	UCC_REMOTE_COUNTERS
+} ucc_mem_hints_t;
+
+typedef struct ucc_mem_map_params {
+    void                    *address;
+    size_t                  len;
+    ucc_mem_hints_t         hints;
+    ucc_mem_constraints_t   constraints;
+} ucc_mem_map_params_t;
+
+typedef struct  ucc_team_oob_coll {
+    int             (*allgather)(void *src_buf, void *recv_buf, size_t size,
+                                 void *allgather_info,  void **request);
+    ucc_status_t    (*req_test)(void *request);
+    ucc_status_t    (*req_free)(void *request);
+    uint32_t 	     participants;
+    void             *coll_info;
+}  ucc_team_oob_coll_t;
+
+typedef enum {
+    UCC_COLLECTIVE_POST_ORDERED = 0,
+    UCC_COLLECTIVE_POST_UNORDERED = 1
+} ucc_post_ordering_t;
+
+typedef enum {
+    UCC_COLLECTIVE_EP_RANGE_CONTIG = 0, 
+    UCC_COLLECTIVE_EP_RANGE_NONCONTIG = 1
+} ucc_ep_range_type_t;
+
+typedef struct ucc_team_params {
+    uint64_t                mask;
+    ucc_post_ordering_t     ordering;
+    uint64_t                outstanding_colls;
+    uint64_t                ep;
+    ucc_ep_range_type_t     ep_range;
+    ucc_coll_sync_type_t    sync_type;
+    ucc_team_oob_coll_t     oob_collective;
+    ucc_mem_map_params_t    mem_params;
```

Comment:
I'm missing topology information. More specifically, I need information about which members of my team are on the same host and which are on the network. Today UCG gets this information from MPI (as a "distance vector"), and I'm not sure how I achieve this using the given API.

Reply:
Yes, the topology is something that we have not handled. In the WG, we decided to push this next version of API. So, it can be handled more comprehensively. Let's touch upon this in the next WG call.